File: src/tag/docs.py
"""Document ingestion — turn documents the model cannot read into text it can.

Shells out to firecrawl/pdf-inspector, discovered on PATH exactly the way ``gh``
already is: used when present, never assumed, and its absence reported as an
absence rather than a failure.

Why a subprocess and not the PyPI binding. One integration then serves BOTH
distributions — the Go harness cannot use the Rust library in-process (no C ABI,
and its wasm build is wasm-bindgen rather than WASI) but it can run the CLI the
npm package ships. Two implementations of "which pages need OCR" would be two
answers to the same question, and that is the incoherence this deliberately
avoids. See docs/prd/PRD-133.

Mirrors ``tag-go/internal/docs``; keep the two in step.
"""
from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from dataclasses import dataclass, field
from pathlib import Path

BINARY_NAME = "pdf-inspector"
ENV_OVERRIDE = "TAG_PDF_INSPECTOR"

#: Bounds one extraction. A path can arrive from a webhook, so the subprocess
#: gets a deadline rather than the caller's patience.
DEFAULT_TIMEOUT = 60.0

#: Caps what an extraction may return. A large PDF extracts to megabytes, and
#: every byte lands in a context window — the size of the result is chosen by
#: the input, so it has to be chosen by us. Exceeding it truncates and SAYS SO.
MAX_MARKDOWN_BYTES = 1 << 20

#: Below this a page is treated as producing no usable text. Deliberately small:
#: a false "needs OCR" costs a wasted OCR call, a false "read fine" costs a
#: silently blank page presented as content.
MIN_CHARS_PER_PAGE = 8

#: Per-page verification budget. One subprocess per page is ~15ms; past this the
#: wall-clock cost stops being worth the certainty.
MAX_VERIFY_PAGES = 64

#: Exact pin for the in-process wheel (PRD-133 Tier 2). A 0.x Rust-built wheel
#: on a fast-moving line is exactly the supply-chain shape the no-ranges policy
#: exists for, so it is pinned here, in the `pdf` extra of pyproject.toml, and in
#: the runtime LAZY_DEPS allowlist — bump all three in lockstep.
PDF_INSPECTOR_PIN = "pdf-inspector==0.2.6"
PDF_INSPECTOR_MODULE = "pdf_inspector"

#: Force a backend for tests / parity checks: "subprocess" | "inprocess".
DOC_BACKEND_ENV = "TAG_DOC_BACKEND"

INSTALL_HINT = (
    "install the in-process engine with `pip install 'tag-agent[pdf]'`, "
    "or the CLI with `npm install -g @firecrawl/pdf-inspector` "
    "(ships a prebuilt binary; no toolchain required), or set "
    f"{ENV_OVERRIDE} to an existing one"
)


class DocumentUnavailable(RuntimeError):
    """The engine is not installed. Distinct so callers can offer the hint."""


class DocumentError(RuntimeError):
    """The engine ran and failed, or returned something unusable."""


class DocumentBadInput(DocumentError):
    """The operator named the wrong path — a missing file or a directory.

    Distinct so the CLI can classify it as a usage error (exit 2), matching the
    Go harness's docs.ErrBadInput and `doc read --help`; a genuine engine
    failure stays a plain DocumentError (exit 1).
    """


@dataclass
class Document:
    path: str
    type: str = ""
    page_count: int = 0
    title: str = ""
    markdown: str = ""
    #: 1-BASED and reconciled — the union of what the engine reported and what
    #: verification found empty. Normalised in one place because the engine's
    #: own APIs disagree about indexing, and an off-by-one in "which page is
    #: unreadable" is the kind of wrong that reads as right.
    pages_needing_ocr: list[int] = field(default_factory=list)
    pages_with_tables: list[int] = field(default_factory=list)
    pages_with_columns: list[int] = field(default_factory=list)
    #: False when any page could not be read, or the text was truncated.
    complete: bool = True
    notes: list[str] = field(default_factory=list)
    engine_ms: int = 0

    def note(self, text: str) -> None:
        if text not in self.notes:
            self.notes.append(text)

    def to_dict(self) -> dict:
        return {
            "path": self.path,
            "type": self.type,
            "page_count": self.page_count,
            "title": self.title,
            "markdown": self.markdown,
            "pages_needing_ocr": self.pages_needing_ocr,
            "pages_with_tables": self.pages_with_tables,
            "pages_with_columns": self.pages_with_columns,
            "complete": self.complete,
            "notes": self.notes,
            "engine_ms": self.engine_ms,
        }


def engine_path() -> str | None:
    """Return the engine's path, or None when it is not installed."""
    override = (os.environ.get(ENV_OVERRIDE) or "").strip()
    if override:
        # An override that does not resolve is an operator mistake worth failing
        # on, not a reason to silently use a different binary from PATH.
        return override if Path(override).is_file() else None
    return shutil.which(BINARY_NAME)


def available() -> bool:
    return engine_path() is not None


def supported(path: str | Path) -> bool:
    """Routing hint by extension. Not a guarantee."""
    return str(path).lower().endswith(".pdf")


def normalize_pages(pages, page_count: int) -> list[int]:
    """Sorted, de-duplicated, 1-based, in-range."""
    out: set[int] = set()
    for p in pages or []:
        if not isinstance(p, int) or p < 1:
            # A 0 can only be a 0-based leak; dropping beats guessing.
            continue
        if page_count > 0 and p > page_count:
            continue
        out.add(p)
    return sorted(out)


def format_pages(pages: list[int]) -> str:
    """Render a page list compactly (1-3, 7)."""
    if not pages:
        return "none"
    pages = sorted(pages)
    parts: list[str] = []
    start = prev = pages[0]

    def flush() -> None:
        if start == prev:
            parts.append(str(start))
        elif prev == start + 1:
            parts.extend([str(start), str(prev)])
        else:
            parts.append(f"{start}-{prev}")

    for p in pages[1:]:
        if p == prev + 1:
            prev = p
            continue
        flush()
        start = prev = p
    flush()
    return ", ".join(parts)


def _run_engine(binary: str, path: str, timeout: float, pages: int | None = None) -> dict:
    args = [binary, path, "--json"]
    if pages is not None:
        args += ["--pages", str(pages)]
    try:
        proc = subprocess.run(args, capture_output=True, text=True, timeout=timeout)
    except subprocess.TimeoutExpired:
        raise DocumentError(f"pdf-inspector timed out after {timeout:g}s") from None
    if proc.returncode != 0:
        msg = (proc.stderr or proc.stdout or "").strip().splitlines()
        raise DocumentError(f"pdf-inspector failed: {msg[0] if msg else proc.returncode}")
    try:
        return json.loads(proc.stdout)
    except json.JSONDecodeError as exc:
        raise DocumentError(f"pdf-inspector returned output that is not JSON: {exc}") from exc


def extract(
    path: str | Path,
    *,
    timeout: float = DEFAULT_TIMEOUT,
    max_bytes: int = MAX_MARKDOWN_BYTES,
    skip_verify: bool = False,
) -> Document:
    """Read *path* and return its text.

    Selects the in-process ``pdf_inspector`` wheel when it is importable, else
    the ``pdf-inspector`` subprocess CLI — both produce the identical Document
    (see :func:`_finalize`). Raises :class:`DocumentUnavailable` when neither is
    present, :class:`DocumentBadInput` for a missing file or a directory, and
    :class:`DocumentError` when the engine runs and fails.
    """
    # A non-positive cap means "use the default", matching the Go harness
    # (opts.MaxBytes <= 0 -> MaxMarkdownBytes). Without this, a negative value
    # became a negative slice markdown[:max_bytes] that silently dropped the
    # TAIL of the document while reporting it as "truncated".
    if max_bytes <= 0:
        max_bytes = MAX_MARKDOWN_BYTES
    backend = _select_backend()
    if backend == "":
        raise DocumentUnavailable(f"no document engine is available — {INSTALL_HINT}")
    p = Path(path)
    if not p.exists():
        raise DocumentBadInput(f"{path}: no such file")
    if p.is_dir():
        raise DocumentBadInput(f"{path} is a directory")

    if backend == "inprocess":
        summary, page_md = _inprocess_backend(str(p), timeout)
    else:
        summary, page_md = _subprocess_backend(str(p), timeout)
    return _finalize(str(p), summary, page_md, max_bytes=max_bytes, skip_verify=skip_verify)


def _select_backend() -> str:
    """Return "inprocess", "subprocess", or "" (nothing available).

    ``TAG_DOC_BACKEND`` forces one, for tests and parity checks. Otherwise the
    in-process wheel wins when it is already importable (no Node required), then
    the CLI on PATH.
    """
    forced = (os.environ.get(DOC_BACKEND_ENV) or "").strip().lower()
    if forced in ("subprocess", "inprocess"):
        return forced
    import importlib.util

    if importlib.util.find_spec(PDF_INSPECTOR_MODULE) is not None:
        return "inprocess"
    if engine_path() is not None:
        return "subprocess"
    return ""


def _subprocess_backend(path: str, timeout: float):
    """The CLI backend: a summary dict + a per-page markdown reader."""
    binary = engine_path()
    if binary is None:
        raise DocumentUnavailable(f"pdf-inspector is not installed — {INSTALL_HINT}")
    summary = _run_engine(binary, path, timeout)

    def page_md(page: int) -> str:  # page is 1-based
        return (_run_engine(binary, path, timeout, pages=page).get("markdown") or "")

    return summary, page_md


def _inprocess_backend(path: str, timeout: float):
    """The pyo3 wheel backend, mapped to the SAME summary shape as the CLI.

    Two things, verified against pdf-inspector==0.2.6, shape this:

    * ``extract_pages_markdown`` takes and reports pages **0-based** (its
      per-page ``needs_ocr`` flag and 0-based ``page`` index are the reliable
      signal), whereas ``process_pdf().pages_needing_ocr`` is **unreliable** — it
      flagged every page of a plainly text-based document. So the OCR list is
      derived from the trustworthy per-page flags, converted to the one 1-based
      convention here, NOT from ``process_pdf``.
    * The per-page markdown is read once and cached, so per-page verification
      costs no extra parse (the subprocess backend re-runs the engine per page).
    """
    mod = _ensure_inprocess()
    r = mod.process_pdf(path)  # doc-level metadata + concatenated markdown
    pages = list(getattr(mod.extract_pages_markdown(path), "pages", []) or [])

    md_by_page: dict[int, str] = {}  # 0-based -> markdown
    ocr_1based: list[int] = []
    for pg in pages:
        idx = int(getattr(pg, "page", 0))  # 0-based
        md_by_page[idx] = getattr(pg, "markdown", "") or ""
        if getattr(pg, "needs_ocr", False):
            ocr_1based.append(idx + 1)

    summary = {
        "pdfType": getattr(r, "pdf_type", "") or "",
        "pageCount": int(getattr(r, "page_count", 0) or 0),
        "title": getattr(r, "title", "") or "",
        "markdown": getattr(r, "markdown", "") or "",
        "pagesWithTables": list(getattr(r, "pages_with_tables", []) or []),
        "pagesWithColumns": list(getattr(r, "pages_with_columns", []) or []),
        "processingTimeMs": int(getattr(r, "processing_time_ms", 0) or 0),
        "pagesNeedingOcr": ocr_1based,
        "hasEncodingIssues": bool(getattr(r, "has_encoding_issues", False)),
        "isComplexLayout": bool(getattr(r, "is_complex_layout", False)),
    }

    def page_md(page: int) -> str:  # page is 1-based; cache is 0-based
        return md_by_page.get(page - 1, "")

    return summary, page_md


def _lazy_installs_allowed() -> bool:
    """Auto-install is OPT-IN. A silent network install of a pinned wheel in a
    supply-chain-cautious codebase must be asked for, not assumed."""
    if _truthy(os.environ.get("HERMES_DISABLE_LAZY_INSTALLS")):
        return False
    return _truthy(os.environ.get("TAG_ALLOW_LAZY_INSTALLS"))


def _truthy(v: str | None) -> bool:
    return (v or "").strip().lower() in ("1", "true", "yes", "on")


def _ensure_inprocess():
    """Import the wheel, lazy-installing the exact pin only when explicitly
    allowed. Absent and not allowed → an honest DocumentUnavailable naming the
    manual install, never a silent pip call."""
    import importlib

    try:
        return importlib.import_module(PDF_INSPECTOR_MODULE)
    except ImportError:
        pass
    if not _lazy_installs_allowed():
        raise DocumentUnavailable(
            f"{PDF_INSPECTOR_MODULE} is not installed — run "
            f"`pip install 'tag-agent[pdf]'` (or set TAG_ALLOW_LAZY_INSTALLS=1 "
            f"to auto-install {PDF_INSPECTOR_PIN})"
        )
    # Exact-pin discipline: only ever install the one pinned spec.
    subprocess.run(
        [sys.executable, "-m", "pip", "install", PDF_INSPECTOR_PIN],
        check=True,
        capture_output=True,
        text=True,
        timeout=300,
    )
    return importlib.import_module(PDF_INSPECTOR_MODULE)


def _finalize(
    path: str,
    summary: dict,
    page_md,
    *,
    max_bytes: int,
    skip_verify: bool,
) -> Document:
    """Turn a backend summary into a Document, applying the honesty logic that
    is identical for both backends: per-page verification, OCR/truncation/empty
    notes, and the 1-place page normalisation."""
    doc = Document(
        path=path,
        type=summary.get("pdfType", ""),
        page_count=int(summary.get("pageCount") or 0),
        title=(summary.get("title") or "").strip(),
        markdown=summary.get("markdown") or "",
        pages_with_tables=summary.get("pagesWithTables") or [],
        pages_with_columns=summary.get("pagesWithColumns") or [],
        engine_ms=int(summary.get("processingTimeMs") or 0),
    )
    doc.pages_needing_ocr = normalize_pages(summary.get("pagesNeedingOcr"), doc.page_count)

    if summary.get("hasEncodingIssues"):
        doc.note("the engine reported encoding issues, so some characters may be wrong")
    if summary.get("isComplexLayout"):
        doc.note("complex layout: reading order may not match the visual order")

    if not skip_verify:
        _verify_pages(doc, page_md)

    if doc.pages_needing_ocr:
        doc.complete = False
        doc.note(
            f"{len(doc.pages_needing_ocr)} of {doc.page_count} page(s) produced no text "
            f"and need OCR: {format_pages(doc.pages_needing_ocr)}"
        )
    if len(doc.markdown.encode("utf-8")) > max_bytes:
        doc.markdown = doc.markdown.encode("utf-8")[:max_bytes].decode("utf-8", "ignore")
        doc.complete = False
        doc.note(
            f"the extracted text exceeded {max_bytes} bytes and was truncated; "
            "this is not the whole document"
        )
    if not doc.markdown.strip() and doc.page_count > 0 and not doc.pages_needing_ocr:
        doc.complete = False
        doc.note(
            "no text was extracted and no page was flagged as needing OCR — the document "
            "could not be read, and this is not evidence that it is empty"
        )
    return doc


def _verify_pages(doc: Document, page_md) -> None:
    """Confirm each unflagged page actually yields text.

    The engine's document-level OCR list is a threshold judgement, not a fact. A
    page that produced no text was not read, whatever the summary said — and
    handing back a document with silently blank pages, labelled complete, is the
    pattern this whole feature exists to avoid. ``page_md`` is the backend's
    1-based per-page markdown reader.
    """
    if doc.page_count <= 0:
        return
    flagged = set(doc.pages_needing_ocr)
    unflagged = [p for p in range(1, doc.page_count + 1) if p not in flagged]
    if not unflagged:
        return

    if len(unflagged) > MAX_VERIFY_PAGES:
        # Density check instead: if the whole document produced less text than
        # one plausible page, the summary is not believable regardless.
        if len(doc.markdown.strip()) < MIN_CHARS_PER_PAGE:
            doc.pages_needing_ocr = normalize_pages(unflagged, doc.page_count)
        doc.note(
            f"per-page verification skipped: {doc.page_count} pages is over the "
            f"{MAX_VERIFY_PAGES}-page budget; the OCR list is the engine's summary "
            "plus a whole-document density check"
        )
        return

    empty: list[int] = []
    for page in unflagged:
        try:
            md = page_md(page)
        except DocumentError as exc:
            doc.note(
                f"per-page verification could not run ({exc}), so the OCR list is the "
                "engine's summary alone and has not been confirmed"
            )
            return
        if len((md or "").strip()) < MIN_CHARS_PER_PAGE:
            empty.append(page)
    if empty:
        doc.pages_needing_ocr = normalize_pages(doc.pages_needing_ocr + empty, doc.page_count)
        doc.note(
            f"verification found {len(empty)} page(s) the engine did not flag but which "
            f"produced no text: {format_pages(empty)}"
        )

File: tag-go/internal/docs/docs.go
// Package docs turns documents the model cannot read into text it can.
//
// It shells out to firecrawl/pdf-inspector, discovered on PATH exactly the way
// `gh` and `git` already are in this harness: used when present, never assumed,
// and its absence is reported as an absence rather than a failure.
//
// Why a subprocess and not a library. pdf-inspector is Rust. Its cdylib is the
// pyo3 extension module -- there is no C ABI to link against -- and its wasm
// build is wasm-bindgen, which wazero cannot run. Linking would also mean CGO,
// and a CGO-free static binary is a commitment this project has made. The npm
// package ships a prebuilt native binary and a CLI, so the process boundary is
// the one door that is actually open. See docs/prd/PRD-133.
//
// The contract this package adds on top of the engine is honesty about what was
// read. The engine's document-level "which pages need OCR" is a threshold
// judgement, not a fact, so Extract cross-checks it: a page the engine implies
// was read must have produced text, and a page that produced none is reported as
// unreadable no matter what the summary said. Handing back a document with
// silently blank pages, labelled complete, would reintroduce the exact pattern
// the read_file refusal was written to remove.
package docs

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

// BinaryName is the executable published by @firecrawl/pdf-inspector.
const BinaryName = "pdf-inspector"

// EnvOverride names an explicit path to the binary, for installs that keep it
// outside PATH (a repo-local node_modules/.bin, for instance).
const EnvOverride = "TAG_PDF_INSPECTOR"

// DefaultTimeout bounds one extraction. Documents are attacker-influenced input
// in the general case -- a webhook can hand TAG a path -- so the subprocess gets
// a deadline rather than the caller's patience.
const DefaultTimeout = 60 * time.Second

// MaxMarkdownBytes caps what an extraction may return.
//
// A large PDF can extract to megabytes of markdown, and every byte of it lands
// in a context window. This is the same class of bound as the SSE accumulators
// and the bash output cap: the size of the result is chosen by the input, so it
// has to be chosen by us instead. Exceeding it truncates and SAYS SO -- silently
// clipping a document and presenting it as whole is the failure mode this
// package exists to avoid.
const MaxMarkdownBytes = 1 << 20 // 1 MiB

// ErrUnavailable is returned when the engine is not installed. It is a distinct
// error so a caller can offer the install hint instead of reporting a failure.
var ErrUnavailable = errors.New("pdf-inspector is not installed")

// ErrBadInput marks operator-supplied path mistakes — a missing file or a
// directory. The CLI classifies these as usage errors (exit 2), matching
// `doc read --help` and the `fix-vuln` sibling; a genuine engine failure is a
// run failure (exit 1) and does not carry this sentinel.
var ErrBadInput = errors.New("bad document input")

// Document is the result of reading one file.
type Document struct {
	Path      string `json:"path"`
	Type      string `json:"type"`       // TextBased | Scanned | ImageBased | Mixed
	PageCount int    `json:"page_count"` // 0 when the engine did not report one
	Title     string `json:"title,omitempty"`
	Markdown  string `json:"markdown"`

	// PagesNeedingOCR is 1-BASED and reconciled: it is the union of what the
	// engine reported and what the verification pass found empty. Indexing is
	// normalised here, in one place, because the engine's own APIs disagree
	// about it and an off-by-one in "which page is unreadable" is the kind of
	// wrong that reads as right.
	PagesNeedingOCR []int `json:"pages_needing_ocr"`

	PagesWithTables  []int `json:"pages_with_tables,omitempty"`
	PagesWithColumns []int `json:"pages_with_columns,omitempty"`

	// Complete is false when any page could not be read, or when the markdown
	// was truncated. A caller must not present an incomplete document as whole.
	Complete bool `json:"complete"`

	// Notes explain every way the result falls short of "the whole document as
	// text". Empty means it does not.
	Notes []string `json:"notes,omitempty"`

	EngineMillis int `json:"engine_ms,omitempty"`
}

// Available reports the engine's path, or false when it is not installed.
func Available() (string, bool) {
	if p := strings.TrimSpace(os.Getenv(EnvOverride)); p != "" {
		if fi, err := os.Stat(p); err == nil && !fi.IsDir() {
			return p, true
		}
		// An override that does not resolve is an operator mistake worth failing
		// on rather than silently falling back to PATH and using a different
		// binary than the one they named.
		return "", false
	}
	p, err := exec.LookPath(BinaryName)
	if err != nil {
		return "", false
	}
	return p, true
}

// InstallHint is the message shown when the engine is absent. It names the one
// command that fixes it.
const InstallHint = "install it with `npm install -g @firecrawl/pdf-inspector` " +
	"(ships a prebuilt binary; no toolchain required), or set " + EnvOverride +
	" to an existing one"

// Supported reports whether this package can be expected to handle the file,
// based on extension alone. It is a routing hint, not a guarantee.
func Supported(path string) bool {
	return strings.EqualFold(filepath.Ext(path), ".pdf")
}

// engineResult mirrors the CLI's --json document payload.
type engineResult struct {
	PDFType          string `json:"pdfType"`
	PageCount        int    `json:"pageCount"`
	ProcessingTimeMs int    `json:"processingTimeMs"`
	PagesNeedingOCR  []int  `json:"pagesNeedingOcr"`
	Confidence       any    `json:"confidence"`
	IsComplexLayout  bool   `json:"isComplexLayout"`
	PagesWithTables  []int  `json:"pagesWithTables"`
	PagesWithColumns []int  `json:"pagesWithColumns"`
	HasEncodingIsses bool   `json:"hasEncodingIssues"`
	Markdown         string `json:"markdown"`
	Title            string `json:"title"`
}

// Options configures Extract.
type Options struct {
	// Timeout bounds the subprocess (0 = DefaultTimeout).
	Timeout time.Duration
	// MaxBytes caps the returned markdown (0 = MaxMarkdownBytes).
	MaxBytes int
	// SkipVerify disables the per-page cross-check. Off by default and only
	// worth setting when the caller has already established the document is
	// text-based; the check is the difference between "the engine said it read
	// this" and "we confirmed it did".
	SkipVerify bool
}

// Extract reads path and returns its text.
func Extract(ctx context.Context, path string, opts Options) (*Document, error) {
	bin, ok := Available()
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnavailable, InstallHint)
	}
	if opts.Timeout <= 0 {
		opts.Timeout = DefaultTimeout
	}
	if opts.MaxBytes <= 0 {
		opts.MaxBytes = MaxMarkdownBytes
	}
	if fi, err := os.Stat(path); err != nil {
		return nil, fmt.Errorf("reading %s: %w (%w)", path, err, ErrBadInput)
	} else if fi.IsDir() {
		return nil, fmt.Errorf("%s is a directory (%w)", path, ErrBadInput)
	}

	res, err := runEngine(ctx, bin, opts.Timeout, path, "--json")
	if err != nil {
		return nil, err
	}

	doc := &Document{
		Path:             path,
		Type:             res.PDFType,
		PageCount:        res.PageCount,
		Title:            strings.TrimSpace(res.Title),
		Markdown:         res.Markdown,
		PagesWithTables:  res.PagesWithTables,
		PagesWithColumns: res.PagesWithColumns,
		EngineMillis:     res.ProcessingTimeMs,
		Complete:         true,
	}
	doc.PagesNeedingOCR = normalizePages(res.PagesNeedingOCR, res.PageCount)
	if doc.PagesNeedingOCR == nil {
		// [] rather than null: this repo's --json contract is an empty list, and
		// a consumer checking `pages_needing_ocr.length` must not hit null.
		doc.PagesNeedingOCR = []int{}
	}

	if res.HasEncodingIsses {
		doc.Note("the engine reported encoding issues, so some characters may be wrong")
	}
	if res.IsComplexLayout {
		doc.Note("complex layout: reading order may not match the visual order")
	}

	// The cross-check. See the package comment: the engine's summary is a
	// judgement, and a page that produced no text was not read whatever the
	// summary says.
	if !opts.SkipVerify {
		if err := doc.verifyPages(ctx, bin, opts.Timeout); err != nil {
			doc.Note("per-page verification could not run (" + err.Error() +
				"), so the OCR list below is the engine's summary alone and has not been confirmed")
		}
	}

	if len(doc.PagesNeedingOCR) > 0 {
		doc.Complete = false
		doc.Note(fmt.Sprintf("%d of %d page(s) produced no text and need OCR: %s",
			len(doc.PagesNeedingOCR), doc.PageCount, formatPages(doc.PagesNeedingOCR)))
	}
	if len(doc.Markdown) > opts.MaxBytes {
		doc.Markdown = truncateAtRune(doc.Markdown, opts.MaxBytes)
		doc.Complete = false
		doc.Note(fmt.Sprintf("the extracted text exceeded %d bytes and was truncated; "+
			"this is not the whole document", opts.MaxBytes))
	}
	if strings.TrimSpace(doc.Markdown) == "" && doc.PageCount > 0 && len(doc.PagesNeedingOCR) == 0 {
		// Nothing extracted, nothing flagged: say so rather than returning an
		// empty string that reads like an empty document.
		doc.Complete = false
		doc.Note("no text was extracted and no page was flagged as needing OCR — " +
			"the document could not be read, and this is not evidence that it is empty")
	}
	return doc, nil
}

// verifyPages confirms that each page the engine did NOT flag actually yields
// text, and folds any that do not into PagesNeedingOCR.
//
// Only unflagged pages are checked, and only when the document is small enough
// that one subprocess per page is affordable. Beyond that a density heuristic
// stands in, and the note says which was used -- an unverified result must not
// look like a verified one.
func (d *Document) verifyPages(ctx context.Context, bin string, timeout time.Duration) error {
	if d.PageCount <= 0 {
		return nil
	}
	flagged := map[int]bool{}
	for _, p := range d.PagesNeedingOCR {
		flagged[p] = true
	}
	unflagged := make([]int, 0, d.PageCount)
	for p := 1; p <= d.PageCount; p++ {
		if !flagged[p] {
			unflagged = append(unflagged, p)
		}
	}
	if len(unflagged) == 0 {
		return nil
	}

	if len(unflagged) > maxVerifyPages {
		// Density check instead: if the whole document produced less text than
		// one plausible page, the summary is not believable regardless.
		if len(strings.TrimSpace(d.Markdown)) < minCharsPerPage {
			d.PagesNeedingOCR = normalizePages(unflagged, d.PageCount)
		}
		d.Note(fmt.Sprintf("per-page verification skipped: %d pages is over the %d-page budget; "+
			"the OCR list is the engine's summary plus a whole-document density check",
			d.PageCount, maxVerifyPages))
		return nil
	}

	var empty []int
	for _, p := range unflagged {
		res, err := runEngine(ctx, bin, timeout, d.Path, "--json", "--pages", strconv.Itoa(p))
		if err != nil {
			return err
		}
		if len(strings.TrimSpace(res.Markdown)) < minCharsPerPage {
			empty = append(empty, p)
		}
	}
	if len(empty) > 0 {
		d.PagesNeedingOCR = normalizePages(append(d.PagesNeedingOCR, empty...), d.PageCount)
		d.Note(fmt.Sprintf("verification found %d page(s) the engine did not flag but which "+
			"produced no text: %s", len(empty), formatPages(empty)))
	}
	return nil
}

const (
	// maxVerifyPages bounds the per-page pass. One subprocess per page is
	// ~15ms; past this the wall-clock cost stops being worth the certainty.
	maxVerifyPages = 64
	// minCharsPerPage is the floor below which a page is treated as producing
	// no usable text. Deliberately small: a near-empty page is legitimate, and
	// the cost of a false "needs OCR" is a wasted OCR call, while the cost of a
	// false "read fine" is a silently blank page presented as content.
	minCharsPerPage = 8
)

// Note appends an explanation, skipping duplicates.
func (d *Document) Note(s string) {
	for _, n := range d.Notes {
		if n == s {
			return
		}
	}
	d.Notes = append(d.Notes, s)
}

func runEngine(ctx context.Context, bin string, timeout time.Duration, args ...string) (*engineResult, error) {
	cctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	cmd := exec.CommandContext(cctx, bin, args...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		if cctx.Err() == context.DeadlineExceeded {
			return nil, fmt.Errorf("pdf-inspector timed out after %s", timeout)
		}
		msg := strings.TrimSpace(stderr.String())
		if msg == "" {
			msg = err.Error()
		}
		return nil, fmt.Errorf("pdf-inspector failed: %s", firstLine(msg))
	}
	var res engineResult
	if err := json.Unmarshal(stdout.Bytes(), &res); err != nil {
		return nil, fmt.Errorf("pdf-inspector returned output that is not JSON: %w", err)
	}
	return &res, nil
}

// normalizePages returns a sorted, de-duplicated, 1-based page list with
// out-of-range entries dropped.
//
// The engine's own APIs disagree about indexing -- the document summary is
// 1-based while the per-page API is 0-based -- so every list is funnelled
// through here. An off-by-one in "which page is unreadable" is the kind of wrong
// that looks right.
func normalizePages(in []int, pageCount int) []int {
	if len(in) == 0 {
		return nil
	}
	seen := map[int]bool{}
	out := make([]int, 0, len(in))
	for _, p := range in {
		if p < 1 {
			continue // a 0 can only be a 0-based leak; dropping beats guessing
		}
		if pageCount > 0 && p > pageCount {
			continue
		}
		if seen[p] {
			continue
		}
		seen[p] = true
		out = append(out, p)
	}
	sort.Ints(out)
	if len(out) == 0 {
		return nil
	}
	return out
}

// formatPages renders a page list compactly (1-3, 7).
func formatPages(pages []int) string {
	if len(pages) == 0 {
		return "none"
	}
	sorted := append([]int(nil), pages...)
	sort.Ints(sorted)
	var parts []string
	start, prev := sorted[0], sorted[0]
	flush := func() {
		switch {
		case start == prev:
			parts = append(parts, strconv.Itoa(start))
		case prev == start+1:
			parts = append(parts, strconv.Itoa(start), strconv.Itoa(prev))
		default:
			parts = append(parts, fmt.Sprintf("%d-%d", start, prev))
		}
	}
	for _, p := range sorted[1:] {
		if p == prev+1 {
			prev = p
			continue
		}
		flush()
		start, prev = p, p
	}
	flush()
	return strings.Join(parts, ", ")
}

// truncateAtRune cuts s to at most n bytes without splitting a character.
func truncateAtRune(s string, n int) string {
	if len(s) <= n {
		return s
	}
	for n > 0 && s[n]&0xc0 == 0x80 {
		n--
	}
	return s[:n]
}

func firstLine(s string) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		return s[:i]
	}
	return s
}
